service/history/queues/executable.go

```diff
@@ -54,3 +66,250 @@ type (
 	// task should always be executed as active or verified as standby
 	TaskFilter func(task tasks.Task) bool
 )
+
+var (
+	schedulerRetryPolicy = common.CreateTaskProcessingRetryPolicy()
+	reschedulePolicy     = common.CreateTaskReschedulePolicy()
+)
+
+const (
+	resubmitMaxAttempts = 10
+)
+
+type (
+	executableImpl struct {
+		tasks.Task
+
+		sync.Mutex
+		state    ctasks.State
+		priority int
+		attempt  int
+
+		executor    Executor
+		scheduler   Scheduler
+		rescheduler Rescheduler
+		timeSource  clock.TimeSource
+
+		loadTime           time.Time
+		userLatency        time.Duration
+		logger             log.Logger
+		scope              metrics.Scope
+		criticalRetryCount dynamicconfig.IntPropertyFn
+
+		queueType     QueueType
+		filter        TaskFilter
+		shouldProcess bool
+	}
+)
+
+func NewExecutable(
+	task tasks.Task,
+	filter TaskFilter,
+	executor Executor,
+	scheduler Scheduler,
+	rescheduler Rescheduler,
+	timeSource clock.TimeSource,
+	logger log.Logger,
+	scope metrics.Scope,
+	criticalRetryCount dynamicconfig.IntPropertyFn,
+	queueType QueueType,
+) Executable {
+	return &executableImpl{
+		Task:               task,
+		state:              ctasks.TaskStatePending,
+		attempt:            1,
+		executor:           executor,
+		scheduler:          scheduler,
+		rescheduler:        rescheduler,
+		timeSource:         timeSource,
+		loadTime:           timeSource.Now(),
+		logger:             tasks.InitializeLogger(task, logger),
+		scope:              scope,
+		queueType:          queueType,
+		criticalRetryCount: criticalRetryCount,
+		filter:             filter,
+	}
+}
+
+func (e *executableImpl) Execute() error {
+	// this filter should also contain the logic for overriding
+	// results from task allocator (force executing some standby task types)
+	e.shouldProcess = e.filter(e.Task)
+	if !e.shouldProcess {
+		return nil
+	}
+
+	ctx := metrics.AddMetricsContext(context.Background())
+	startTime := e.timeSource.Now()
+	err := e.executor.Execute(ctx, e)
+	var userLatency time.Duration
+	if duration, ok := metrics.ContextCounterGet(ctx, metrics.HistoryWorkflowExecutionCacheLatency); ok {
+		userLatency = time.Duration(duration)
+	}
+	e.userLatency += userLatency
```

Comment:
so this e.userLatency will accumulate during retry?

Reply:
Yes, `e.userLatency` is used to calculate `taskNoUserLatency` and `taskNoUserQueueLatency` when acking the task. Both latencies are across multiple attempts/retries. This is the same as the existing impl in `taskProcessor.go`

However, I believe existing logic `taskProcessor.go` is not calculating the `taskProcessingLatency` correctly, as it's calculated by subtracting `e.userlatency` which is across attempts from the execution latency for one attempt.